add mmctl ability to export only users

svelle opened the pull request:
We currently only support a very limited set of combination to migrate user authentication methods often resulting in the need to have to either modify the db, which is very bad, or having to migrate to email/pass first and then go to the desired auth method and even that doesn't always work.

This PR adds the ability for mmctl to export *only* a server's users. This way the admin can then go in and edit that file to use the new auth method or edit any other attribute and reimport it, not having to export all posts, channels, etc. will be massive time and resource saver and also not cause any confusion in terms of data needing to be removed after the export is done.

svelle committed:
add mmctl ability to export only users

svelle committed:
add user-only export tests

svelle committed:
add mmctl docs

hanzei reviewed (commented):
Can you please take a look at the ci failures? Happy to review the PR in depth once they are resolved.

svelle commented:
@mattermost/cloud fix the linting problems

claude[bot] committed:
Add documentation comment for UsersOnly field

This addresses the linting issue where the public UsersOnly field
was missing a documentation comment.

Co-authored-by: svelle <svelle@users.noreply.github.com>

svelle committed:
fix golangci-lint errors

svelle committed:
Merge branch 'master' into user-export

svelle commented:
> Can you please take a look at the ci failures? Happy to review the PR in depth once they are resolved.

@hanzei CI problems are resolved and no issues left (besides the flaky tests)

hanzei commented on server/channels/app/export_test.go:
```diff
@@ -1771,3 +1771,148 @@ func TestExportDeactivatedUserDMs(t *testing.T) {
+	require.Nil(t, appErr)
+
+	user2, appErr := th.App.CreateUser(th.Context, &model.User{
+		Email:    "user2@example.com",
+		Username: "user2",
+		Password: "password123",
+	})
+	require.Nil(t, appErr)
+
+	// Add users to teams and channels
+	_, appErr = th.App.JoinUserToTeam(th.Context, team, user1, "")
+	require.Nil(t, appErr)
+	_, appErr = th.App.JoinUserToTeam(th.Context, team, user2, "")
+	require.Nil(t, appErr)
+
+	_, appErr = th.App.AddUserToChannel(th.Context, user1, channel, false)
+	require.Nil(t, appErr)
+	_, appErr = th.App.AddUserToChannel(th.Context, user2, channel, false)
+	require.Nil(t, appErr)
+
+	// Create some posts (these should NOT be exported in users-only mode)
+	post1, appErr := th.App.CreatePost(th.Context, &model.Post{
+		UserId:    user1.Id,
+		ChannelId: channel.Id,
+		Message:   "Test post 1",
+	}, channel, model.CreatePostFlags{SetOnline: true})
+	require.Nil(t, appErr)
+	require.NotNil(t, post1)
+
+	// Create user preferences
+	preferences := model.Preferences{
+		{
+			UserId:   user1.Id,
+			Category: model.PreferenceCategoryTheme,
+			Name:     "",
+			Value:    `{"type":"custom"}`,
+		},
+	}
+	appErr = th.App.UpdatePreferences(th.Context, user1.Id, preferences)
+	require.Nil(t, appErr)
```
Can we just use the data created via `InitBasic`?

hanzei commented on server/channels/app/export_test.go:
```diff
@@ -1771,3 +1771,148 @@ func TestExportDeactivatedUserDMs(t *testing.T) {
+	appErr = th.App.BulkExport(th.Context, &b, "", nil, opts)
+	require.Nil(t, appErr)
+
+	// Parse the export and verify contents
+	scanner := bufio.NewScanner(&b)
+	hasUsers := false
+	hasTeams := false
+	hasPosts := false
+	hasChannels := false
+	userCount := 0
+
+	for scanner.Scan() {
+		var line imports.LineImportData
+		err := json.Unmarshal(scanner.Bytes(), &line)
+		require.NoError(t, err)
+
+		switch line.Type {
+		case "user":
+			hasUsers = true
+			userCount++
+			// Verify user data is present
+			require.NotNil(t, line.User)
+			require.NotEmpty(t, line.User.Username)
+			require.NotEmpty(t, line.User.Email)
+		case "team":
+			hasTeams = true
+		case "channel":
+			hasChannels = true
+		case "post":
+			hasPosts = true
+		}
+	}
```
Can we call `app.BulkImport` instead? The custom scanner feels a bit hacky.

hanzei commented on server/cmd/mmctl/commands/export.go:
```diff
@@ -134,24 +135,36 @@ func init() {
-	if !excludeRolesAndSchemes {
-		data["include_roles_and_schemes"] = "true"
-	}
+		// For users-only export, only profile pictures option is relevant
```
Actually, I think we should error if other flags are used. They have no effect when using `users-only`.

hanzei commented on server/cmd/mmctl/commands/export_test.go:
```diff
@@ -87,6 +87,85 @@ func (s *MmctlUnitTestSuite) TestExportCreateCmdF() {
+			Return(mockJob, &model.Response{}, nil).
+			Times(1)
+
+		cmd := &cobra.Command{}
+		cmd.Flags().Bool("users-only", true, "")
+		cmd.Flags().Bool("include-profile-pictures", true, "")
```
I dislike that we are duplicating the implementation here. Can we use a similar approach as in https://github.com/mattermost/mattermost/blob/05585e5388b708e06a71c18a1ea64bc0f16087c2/server/cmd/mmctl/commands/ldap.go#L23-L39?

hanzei reviewed (changes requested):
Nice work! Looking forward to the feature :tada:

claude[bot] committed:
Address review comments: refactor users-only export feature

- Change flag from --users-only to --only-users for consistency
- Add error checking for incompatible flags with --only-users
- Refactor export.go: move users-only logic to separate helper methods
- Update tests to use InitBasic data instead of creating new users
- Add BulkImport validation and export file validation to tests
- Update mmctl tests to properly test error cases

Co-authored-by: svelle <svelle@users.noreply.github.com>

hanzei commented:
Well, Claude got 80% right. Still, manual intervention is needed.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!
